### PR TITLE
Audit log enhancements for Kubectl access path

### PR DESCRIPTION
**Pull request by prachidamle**
https://github.com/rancher/rancher/issues/33126

This fix is to support audit logging from kubectl access which use kubeconfig tokens. In case of kubeconfig, the AuthProvider is always set to "local", the userPrincipal is not found and is missing information. Hence the audit log data is not seen. 
As a fix, the kubeconfig token is now fixed to carry the correct authProvider going forward.

Another part of the fix is instead reading extraInfo from user's login token, this is now read from userattributes where the extra info is saved after user logs in. This information can be found using the userId extracted from any token.
The userAttributes.Extra is also saved to ClusterUserAttributes.Extra field.

The process of finding the extra attributes for the user will be:
- Check if  extra attributes are found in userattributes for the token's provider
- If the token's provider is empty or is local, then add all extraAttributes into the Audit entry to cover all IDP values 
- Fallback is read from token if found, or otherwise read from the user object

### CODE REVIEW

File: pkg/auth/requests/authenticate.go
Comment (prachidamle) [inline pkg/auth/requests/authenticate.go:163]: For kubectl access via kubeconfig token, the auth provider is always set as "local" in that token. So we gather extraInfo of all external authProviders and log that to audit-log. The user info for "local" auth is already set in the audit logging, what we desire is having the external IDP (IdentityProvider) values. For UI access, the session tokens always have the correct authProvider configured, so only that provider's extra-info will be passed down.

**Comment by prachidamle** [comment]
Discussed with @dramich and @aiyengar2 and following changes/alternatives were identified:
1) This solution is saving extraInfo from user's login token into userAttribute CR and reading it back by using the token's userId. But we need to additionally lookup principals of the user from the token or from userId if extraInfo is not found in userAttribute  - this is necessary to take care of the upgrade case where old tokens/old userAttribute will not have the extraInfo
2) We also should ideally fix kubeconfig token to carry the real authProvider and not set "local"; this will let us use the correct principalId of that provider instead of us adding a list of all principalIds to audit log

File: pkg/apis/management.cattle.io/v3/authn_types.go
Comment (snasovich) [inline pkg/apis/management.cattle.io/v3/authn_types.go:95]: Wondering if `Extra` name with long type like `map[string]map[string][]string` is too generic for something that looks to me like extra fields per provider? Just looking at field name and type it's not clear that the key of the map is the authn provider name.
It was much more clear when type was `map[string][]string` but addition of extra layer makes it more obscure I think.
Or are there limitations for adding more types here?

File: pkg/auth/providers/activedirectory/activedirectory_provider.go
Comment (snasovich) [inline pkg/auth/providers/activedirectory/activedirectory_provider.go:234]: These checks don't seem to be related to the issue addressed. Is it just an extra check added to not add empty data or there is more subtle reason?
Is it pretty common to have one of or both of `userPrincipal.Name`/`userPrincipal.LoginName` empty?

File: pkg/auth/providers/activedirectory/activedirectory_provider.go
Comment (snasovich) [inline pkg/auth/providers/activedirectory/activedirectory_provider.go:235]: Somewhat of a nitpick, but it could be a good idea to introduce consts for repeating string literals like `"principalid"` and `"username"`.

File: pkg/auth/requests/authenticate.go
Comment (snasovich) [inline pkg/auth/requests/authenticate.go:160]: This sounds more like Debug level message rather than Info. Same for the other `Infof` below.

File: pkg/controllers/managementuser/clusterauthtoken/user_attribute.go
Comment (snasovich) [inline pkg/controllers/managementuser/clusterauthtoken/user_attribute.go:76]: Could we just add `Extra` to `userAttributeCompare` instead? I believe `reflect.DeepEqual` handles maps OK.

File: pkg/apis/management.cattle.io/v3/authn_types.go
Comment (prachidamle) [reply pkg/apis/management.cattle.io/v3/authn_types.go:95]: @snasovich that map[string][]string is as defined by Kubernetes struct to carry userInfo https://pkg.go.dev/k8s.io/apiserver@v0.20.2/pkg/authentication/user#DefaultInfo

And we need to carry this per authProvider hence another map of provider -> defaultInfo so that we do not have to translate it from some custom type when sending it downstream

File: pkg/auth/providers/activedirectory/activedirectory_provider.go
Comment (prachidamle) [reply pkg/auth/providers/activedirectory/activedirectory_provider.go:234]: yes that data could be empty, especially for kubeconfig tokens. This is to add keys only with non-empty values so that we can detect if a fallback on reading the missing data is needed in here https://github.com/rancher/rancher/pull/33268/files#diff-ccde1c52dd5ce3e2e7e9d8c62b6a7a60eb7f80d86942e367c19586fce6b70c63R191-R196

File: pkg/auth/providers/activedirectory/activedirectory_provider.go
Comment (prachidamle) [reply pkg/auth/providers/activedirectory/activedirectory_provider.go:235]: sure

File: pkg/apis/management.cattle.io/v3/authn_types.go
Comment (snasovich) [reply pkg/apis/management.cattle.io/v3/authn_types.go:95]: @prachidamle , thank you for the explanation. `map[string][]string` makes sense since it's what referenced in k8s struct as you linked. I still think the name `Extra` doesn't make it obvious what is used as the key, so maybe we can rename this field to something like `ExtraByProvider` or `ProviderExtra` - provided we can map it to Extra of type `map[string][]string` that is actually logged as "extra" field on audit log entry (I'm not sure what type is actually used to produce audit log entries)?

File: pkg/user/manager.go
Comment (dramich) [inline pkg/user/manager.go:19]: I think we should drop this type and have a clusterName param, seems cleaner vs wrapping the object above when needing to call `EnsureClusterToken`

File: pkg/user/manager.go
Comment (dramich) [inline pkg/user/manager.go:28]:
```suggestion
	EnsureClusterToken(clusterName string, input ClusterTokenInput) (string, error)
```

File: pkg/user/manager.go
Comment (dramich) [inline pkg/user/manager.go:9]: Can we move this to `pkg/auth/providers/common/usermanager.go`? This file is just for the interface as there used to be cyclical imports

File: pkg/auth/requests/authenticate.go
Comment (dramich) [inline pkg/auth/requests/authenticate.go:179]: How does this handle upgrades where all tokens were set to `local`?

File: pkg/auth/requests/authenticate.go
Comment (dramich) [inline pkg/auth/requests/authenticate.go:181]:
```suggestion
			if extraInfo, ok := attribs.Extra[token.AuthProvider]; ok {
			    return extraInfo
			}
```

File: pkg/auth/requests/authenticate.go
Comment (dramich) [inline pkg/auth/requests/authenticate.go:177]: Can return here after the loop since this never sets readFromToken

File: pkg/auth/requests/authenticate.go
Comment (dramich) [inline pkg/auth/requests/authenticate.go:167]: This won't be needed with the returns

File: pkg/auth/requests/authenticate.go
Comment (dramich) [inline pkg/auth/requests/authenticate.go:196]: This should end up being fallback logic and not need the if check

File: pkg/apis/management.cattle.io/v3/authn_types.go
Comment (dramich) [reply pkg/apis/management.cattle.io/v3/authn_types.go:95]: I agree with changing the name as the other fields called `Extra` are for an individual provider, `ExtraByProvider` sounds fine. Can a comment be added as well explaining the field? That type in general is intimidating so having an explanation and example would help.

Comment (prachidamle) [reply pkg/apis/management.cattle.io/v3/authn_types.go:95]: Done

File: pkg/user/manager.go
Comment (prachidamle) [reply pkg/user/manager.go:9]: Tried that but it still creates a cyclic import
package github.com/rancher/rancher
	imports github.com/rancher/rancher/pkg/data/management
	imports github.com/rancher/rancher/pkg/agent/clean
	imports github.com/rancher/rancher/pkg/controllers/management/auth
	imports github.com/rancher/rancher/pkg/auth/providerrefresh
	imports github.com/rancher/rancher/pkg/auth/providers
	imports github.com/rancher/rancher/pkg/auth/api/secrets
	imports github.com/rancher/rancher/pkg/auth/providers/common
	imports github.com/rancher/rancher/pkg/auth/tokens
	imports github.com/rancher/rancher/pkg/types/config
	imports github.com/rancher/rancher/pkg/user
	imports github.com/rancher/rancher/pkg/auth/providers/common: import cycle not allowed

File: pkg/auth/requests/authenticate.go
Comment (prachidamle) [reply pkg/auth/requests/authenticate.go:179]: So yeah I think the original idea where if "local" we write down all provider's  info, that will handle the case where all kubeconfig tokens were set to "local". Also for this feature, external IDP value is the goal, not the local values. But we can set all. Thoughts?

File: pkg/controllers/managementuser/clusterauthtoken/user_attribute.go
Comment (snasovich) [reply pkg/controllers/managementuser/clusterauthtoken/user_attribute.go:79]: @prachidamle , it doesn't seem to be resolved yet.

**Review by snasovich — APPROVED** [review]
There is only one trivial comment that doesn't seem to be resolved yet, but otherwise LGTM. Still need @dramich 's approval.

File: pkg/controllers/managementuser/clusterauthtoken/user_attribute.go
Comment (prachidamle) [reply pkg/controllers/managementuser/clusterauthtoken/user_attribute.go:79]: @snasovich what do you mean by old code?

Comment (snasovich) [reply pkg/controllers/managementuser/clusterauthtoken/user_attribute.go:79]: I mean the one before the change. Existing `master` has one line instead of these 5:
```go
return groups, reflect.DeepEqual(current, old)
```

**Comment by snasovich** [comment]
@prachidamle , even if this is approved, please do not merge it to `master` yet as it was decided this should not get into 2.6.0  release. So, until branches are created for release/v2.6 (and likely release/v2.6) we should't put it in `master`.

**Comment by prachidamle** [comment]
@snasovich I have reopened this against release/v2.6, added you as well fro review since you had reviewed and approved it earlier.
Closing this one.

**Pull request closed without merging**
